### PR TITLE
Extend supported types by QLinearMatMul (float16, float 8 types)

### PR DESCRIPTION
**Pull request by xadupre**
### Description
QLinearMatMul is used on quantized types. This PR extends the list of supported quantized types to float 8 types and the list of supported inputs types to float16, bfloat16.

### CODE REVIEW

File: onnx/defs/math/utils.cc [hunk 1]
```diff
@@ -69,4 +70,115 @@ from the back. Accepted range is [-r, r-1] where r = rank(input).
     });
   };
 }
+
+void matmulShapeInference(ONNX_NAMESPACE::InferenceContext& ctx, int input1Idx, int input2Idx) {
```
Comment (justinchuby) [inline onnx/defs/math/utils.cc:74]:
```suggestion
void MatmulShapeInference(ONNX_NAMESPACE::InferenceContext& ctx, int input1Idx, int input2Idx) {
```

nit: PascalCase for functions?

File: onnx/defs/math/utils.cc [hunk 2]
```diff
@@ -69,4 +70,115 @@ from the back. Accepted range is [-r, r-1] where r = rank(input).
...
+    }
+    if (shape1.dim_size() != 1) {
+      *resultShape.add_dim() = shapeR.dim(shapeR.dim_size() - 1);
+    }
+  }
+
+  *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape() = resultShape;
+}
+
+void qlinear_matmul_shape_inference(ONNX_NAMESPACE::InferenceContext& ctx) {
```
Comment (justinchuby) [inline onnx/defs/math/utils.cc:142]: PascalCase for consistency

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -2020,43 +1922,35 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+        .TypeConstraint("TS", {"tensor(float)", "tensor(float16)", "tensor(bfloat16)"}, "Constrain scales.")
         .TypeConstraint(
             "T1",
-            {"tensor(int8)", "tensor(uint8)"},
+            {"tensor(int8)",
+             "tensor(uint8)",
+             "tensor(float8e4m3fn)",
+             "tensor(float8e4m3fnuz)",
+             "tensor(float8e5m2)",
+             "tensor(float8e5m2fnuz)"},
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:1933]: Suggest changing the line below to "The type of input a and its zeropoint.". It is no longer just 8-bit integer. But no point in repeating the list of all types, I think it is simpler to use the description to say what inputs type is being described. Same for input b and output.

Comment (xadupre) [reply onnx/defs/math/defs.cc:1933]: done

File: onnx/defs/math/utils.cc (hunk 1 above)
Comment (justinchuby) [reply onnx/defs/math/utils.cc:74]: I haven’t spent long enough time in the C++ codebase but I would suggest https://google.github.io/styleguide/cppguide.html#Function_Names being generally a good guide to follow. More importantly I would just strive for local consistency.

Comment (gramalingam) [reply onnx/defs/math/utils.cc:74]: Correct. The existing code is a mix of various conventions. I have been recently trying to stick to the Google styleguide, which is what onnxruntime also uses. We should document this somewhere.

**Review by gramalingam — APPROVED** [review]
At some point, we should try to turn this into a function "Dequantize => MatMul => Quantize". But fine if that is done separately. (There may be some questions about precision of intermediate values etc. there.)

**Comment by WilliamTambellini** [comment]
Tks @xadupre  Does it mean int4 is not yet supported ?

**Comment by xadupre** [comment]
Int4 is not defined yet in onnx. That would be the first step before adding it to the list of supported types. Mayne it would be worth discussing it during one of the sig meeting.

**Review by gramalingam — CHANGES_REQUESTED** [review]
(As per offline discussion, Xavier suggests holding off this PR changes until better float 8 quantization support is available. Adding this comment to avoid accidental merge.)

**Comment by justinchuby** [comment]
Moved to 1.16

**Comment by WilliamTambellini** [comment]
That s unfortunate. What about reducing the scope and at least add uint8/int8 (no float8)?
Fast int8 is now already available in most (4th gen) CPUs, and most recent nvidia gpus.

File: onnx/version_converter/convert.h [hunk 4]
```diff
@@ -606,6 +606,12 @@ class DefaultVersionConverter : public BaseVersionConverter {
         std::make_unique<TypeRestriction>("ReduceMax", OpSetID(20), OpSetID(19), reduce_min_max_18_unallowed_types));
     registerAdapter(
         std::make_unique<TypeRestriction>("ReduceMin", OpSetID(20), OpSetID(19), reduce_min_max_18_unallowed_types));
+
+    /******** 20 -> 21 ********/
+    registerAdapter(std::make_unique<CompatibleAdapter>("QLinearMatMul", OpSetID(20), OpSetID(21)));
+
+    /******** 21 -> 20 ********/
+    registerAdapter(std::make_unique<CompatibleAdapter>("QLinearMatMul", OpSetID(21), OpSetID(20)));
```
Comment (gramalingam) [inline onnx/version_converter/convert.h:614]: This should restrict the conversion to the types supported in earlier version. Eg., see [here](https://github.com/onnx/onnx/blob/1784aeee3a8a9c819e81e516048475dec1d545fd/onnx/version_converter/convert.h#L602)

Comment (gramalingam) [reply onnx/version_converter/convert.h:614]: Hmmm .. except that the old linked example above has the version-numbers swapped (which is a typo/bug that should be fixed, I guess).

Comment (xadupre) [reply onnx/version_converter/convert.h]: done

**Review by justinchuby — APPROVED** [review]

File: onnx/defs/math/utils.cc (hunk 2 above)
Comment (xadupre) [reply onnx/defs/math/utils.cc]: done

**Review by gramalingam — APPROVED** [review]

**Pull request merged**